itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcTimeout.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hive.jdbc;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hive.jdbc.miniHS2.MiniHS2;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.sql.DriverManager;
+import java.sql.Statement;
+import java.util.HashMap;
+
+import static org.junit.Assert.assertEquals;
+
+public class TestJdbcTimeout {
+
+  private static MiniHS2 miniHS2 = null;
+
+  @BeforeClass
+  public static void beforeTest() throws Exception {
+    Class.forName(MiniHS2.getJdbcDriverName());
+    HiveConf conf = new HiveConf();
+    conf.setVar(HiveConf.ConfVars.HIVE_LOCK_MANAGER, "org.apache.hadoop.hive.ql.lockmgr.EmbeddedLockManager");
+    miniHS2 = new MiniHS2.Builder().withConf(conf).build();
+    miniHS2.start(new HashMap<>());
+  }
+
+  @AfterClass
+  public static void afterTest() throws Exception {
+    if (miniHS2 != null && miniHS2.isStarted()) {
+      miniHS2.stop();
+    }
+  }
+
+  @Test
+  public void testDriverManagerLoginTimeout() throws Exception {
+    DriverManager.setLoginTimeout(10);
+    String url = miniHS2.getJdbcURL("default");
+    try (HiveConnection conn = (HiveConnection) DriverManager.getConnection(url)) {
+      assertEquals(10000, conn.getLoginTimeout());
+      assertEquals(0, conn.getSocketTimeout());
+    }
+  }
+
+  @Test
+  public void testParameterLoginTimeout() throws Exception {
+    DriverManager.setLoginTimeout(10);
+    String url = miniHS2.getJdbcURL("default", "loginTimeout=20000");
+    try (HiveConnection conn = (HiveConnection) DriverManager.getConnection(url)) {
+      assertEquals(20000, conn.getLoginTimeout());
+      assertEquals(0, conn.getSocketTimeout());
+    }
+  }
+
+  @Test
+  public void testParameterSocketTimeout() throws Exception {
+    DriverManager.setLoginTimeout(0);
+    String url = miniHS2.getJdbcURL("default", "socketTimeout=10000");
+    try (HiveConnection conn = (HiveConnection) DriverManager.getConnection(url)) {
+      assertEquals(0, conn.getLoginTimeout());
+      assertEquals(10000, conn.getSocketTimeout());
+    }
+  }
+
+  @Test
+  public void testLoginTimeoutDoesNotAffectSocketTime() throws Exception {
```

Comment:
This test does not fail before the PR

Reply:
It should fail before HIVE-12371, HIVE-12371 changes behavior but does not provide test

Reply:
Actually, I don't know how to test `loginTimeout`, does Hive have something like "session init sql"?

Reply:
Could you use simple jdbc connection?
`miniHS2.getJdbcURL()` could provide the jdbc url to connect to the running HS2 server

Reply:
Then what does Spring Boot use? How do we get exceptions with Spring?

Reply:
Couldn't we do the same in our tests?

Reply:
I think I have already done it in this case...

Reply:
Because of HIVE-12371 PR https://github.com/apache/hive/pull/1611

Reply:
> Before [HIVE-12371](https://issues.apache.org/jira/browse/HIVE-12371), the Hive JDBC Driver uses DriverManager#loginTimeout as both connectTimeout and socketTimeout, which usually cause socket timeout exception to users who use Hive JDBC Driver in Spring Boot project, because Spring Boot will setLoginTimeout to 30s (default value).

> HIVE-12371 introduced a new parameter socketTimeout, and does not care about DriverManager#loginTimeout anymore

So by my understanding the PR is aimed for separating the loginTimeout and the socketTimeout. So there should be a situation where having these different configurations would allow us to connect to HS2 after the PR, but not before the PR.

What would be this situation?

Thanks,
Peter

Reply:
> What would be this situation?

I do not know how to explain it, because I think it's fundamental two different things.

MySQL[1], PostgreSQL[2], Thrift, Socket and everything distinguish them

[1] https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-connp-props-connection-authentication.html
[2] https://jdbc.postgresql.org/documentation/head/connect.html

Reply:
Emm... it's not suitable to map `loginTimeout` to `connectTimeout`, how about providing `connectTimeout` and `socketTimeout` and ignore `DriverManager#loginTimeout` ?

Reply:
New idea

```
loginTimeout = parameter.loginTimeout ? parameter.loginTimeout : DriverManager#loginTimeout
connectTimeout = parameter.connectTimeout
socketTimeout = parameter.connectTimeout
```

Before openSession
```
TSocket(connectTimeout = connectTimeout, scoketTimeout = loginTimeout)
THttpClient(connectTimeout = connectTimeout, readTimeout = loginTimeout)
```

After openSession

```
TSocket.setScoketTimeout(socketTimeout)
THttpClient.setReadTimeout(socketTimeout)
```

Does it make sense?

Reply:
Is it possible to change the socketTimeout after opening the socket?
Do we really want to have 2 different timeouts? Maybe we just have to deprecate the loginTimeout

Reply:
> Is it possible to change the socketTimeout after opening the socket?

Yes.

At least I think it's important to have `connectTimeout` and `socketTimeout`, and nice to have `loginTimeout`.

Reply:
yeah

Reply:
That sounds good!

Reply:
How will we test it?

Reply:
I don't know how to test `connectTimeout`, but for `socketTimeout`, I think we just need to rename `loginTimeout` to `connectTime` in this method. 

Then we do not care about `loginTimeout` at all?

Reply:
I think we should honor the `DriverManager.loginTimeout` as a socketTimeout and possible conectionTimeout if they are not set otherwise, just remember that the `DriverManager.loginTimeout` supposed to be in seconds.

As for the testing I have found this:
https://stackoverflow.com/questions/100841/artificially-create-a-connection-timeout-error

> Connect to a non-routable IP address, such as 10.255.255.1

Reply:
> I think we should honor the `DriverManager.loginTimeout` as a `socketTimeout` and possible `connectionTimeout` if they are not set otherwise.

I think it make sense to use `DriverManager.loginTimeout` as `connectTimeout` but not `socketTimeout`. Comparing login or connect, a query may take hours, that is why users report socket timeout exceptions.

Reply:
> As for the testing I have found this:
> https://stackoverflow.com/questions/100841/artificially-create-a-connection-timeout-error

Not sure if it's suitable, from my understanding, it can be used to test socket timeout but not connect timeout. Because the server must accept the connection and establish a TCP connection before reading a parameter(sleep=6000) passed by HTTP protocol.

Reply:
> I think it make sense to use DriverManager.loginTimeout as connectTimeout but not socketTimeout. Comparing login or connect, a query may take hours, that is why users report socket timeout exceptions.

Was this loginTimeout used anywhere else? Do we need to keep backward compatibility?

> >As for the testing I have found this:
> >https://stackoverflow.com/questions/100841/artificially-create-a-connection-timeout-error
>
> Not sure if it's suitable, from my understanding, it can be used to test socket timeout but not connect timeout. Because the server must accept the connection and establish a TCP connection before reading a parameter(sleep=6000) passed by HTTP protocol.

We can have different tests. One for the connectTimeout, and one for the socketTimeout, and set the other to a value which does not interfere with the test.

Reply:
> Was this `loginTimeout` used anywhere else?

I think no, the IDEA index indicates it is not used by others.

> Do we need to keep backward compatibility?

I think not necessary, the old implementation does not right. HIVE-12371 has already broken it one time.

To clarify I think we have 3 approaches for improving the "*timeout" stuff

1. Introduce new JDBC parameter `connectTimeout` w/ exsiting `socketTime`, ignore `DriverManager.loginTimeout`
2. Same as 1, but use `DriverManager.loginTimeout` as `connectTimeout` if `connectTimeout` absent.
3. Introduce new JDBC parameter `loginTimeout` and `connectTimeout` w/s exsiting `socketTime`, use `loginTimeout` as `socketTimeout` for openSession request, and reset `socketTimeout` if openSession successful.

Which one do you prefer?

Reply:
I would go with the 1st approach

Reply:
Thanks for the confirmation, will implement it in the next few days.